MVP: new user/app mode (UI changes only)

lhao03 opened the pull request:
The permissions toggling (UI changes only) is ready to be merged into `interns/dev`.

Demo: 
![mvp-rsc](https://user-images.githubusercontent.com/55033656/124340823-d7e51e80-db6c-11eb-9faa-108bf78b431d.gif)

Complete feature:
- ability to switch to user mode or app mode in the three dots menu
- when you switch between user and app mode, it shows beside the user's name 
- when you check the permissions, they change (delegated vs. rsc)
- you can consent to permissions when you're a delegated user but not when you're in app mode

lhao03 committed:
add packagelockjson to gitignore

lhao03 committed:
remove package-lock.json from git history

lhao03 committed:
add hardcoded switch statement for scopes for permissions

lhao03 committed:
got hardcoded perms working (copied class)

lhao03 committed:
finished hard coded perms tabs

lhao03 committed:
change to delegated work

Commit:
hardcoded the apps in the sidebar with a new tab for RSC

Commit:
scroll bar

Commit:
adding a toggle UI

lhao03 committed:
added constants

lhao03 committed:
package-lock.json

lhao03 committed:
Merge branch 'interns/lhao03/37710-create-a-hardcoded-component' into interns/lhao03/38011-create-the-ui-components-for-permissions

Commit:
dissappears when minimized and appears when authenticated

Commit:
added two versions of the toggle

Commit:
removed the toggle in AppTitle

Commit:
Revert "removed the toggle in AppTitle"

This reverts commit 5e96bd8ef8614ec2bb02cd497e2498e39652bb23.

Commit:
bug fix where the toggle disappeared when the window minimised

lhao03 committed:
redux setup

lhao03 committed:
redux setup

lhao03 committed:
Merge branch 'interns/gracet9182/38013-create-toggles-rsc' into interns/lhao03/38011-create-the-ui-components-for-permissions

lhao03 committed:
constants

lhao03 committed:
attached redux to button

lhao03 committed:
remove hardcoded app perms

lhao03 committed:
added as person as app

lhao03 committed:
#38011 finished implementing perms change

lhao03 committed:
AB#38011 clean up log statements

lhao03 committed:
remove auth for gif

lhao03 committed:
add as user as app

lhao03 committed:
typo

lhao03 committed:
AB#38011 add auth back

lhao03 committed:
remove toggle code

lhao03 committed:
clean up unused imports + files

lhao03 committed:
Merge branch 'interns/lhao03/38011-create-the-ui-components-for-permissions' into interns/dev

lhao03 committed:
fix linting errors

lhao03 committed:
remove spaces

lhao03 committed:
remove need to be authenticated

lhao03 committed:
add tests

lhao03 committed:
add aria label

lhao03 committed:
Interns/lhao03/38035 filter rsc permissions (#9)

* add constants for rsc endings

* add filter functionality

lhao03 committed:
fix lint error

lhao03 committed:
remove reducerr in index that was used for testing

jerryxihe committed:
Moved the option for switching between user and app mode into the drop-down

lhao03 committed:
fixed based on reviewers comments

lhao03 committed:
change if statement and add back delegatedPersonal

lhao03 committed:
add new line and remove useless file

lhao03 committed:
removed misc css

jerryxihe committed:
Removed temporary mode-switching button

jerryxihe committed:
Removed Consent buttons under Modify permissions tab when using app mode

lhao03 committed:
Merge pull request #7 from LokiLabs/interns/lhao03/38235-resolve-merge-conflict-for-perms-ui

AB#38235 merge switching perms

jerryxihe committed:
Removed Admin consent required column under Modify permissions tab when using app mode

jerryxihe committed:
Moved user/app mode identifier from AppTitle to Profile

jerryxihe committed:
Removed unused user/app mode code

jerryxihe committed:
Cleaned up permissions tab conditional formatting

jerryxihe committed:
More descriptive permission type boolean

lhao03 committed:
display delegated / application permissions in panel

jerryxihe committed:
Merge pull request #12 from LokiLabs/interns/jerryxihe/38033-MVP-user-app-mode-UI

AB#38033 MVP user/app mode UI

ayubun commented on src/app/views/authentication/profile/Profile.tsx:
```diff
@@ -37,6 +36,7 @@ export class Profile extends Component<IProfileProps, IProfileState> {
       : null;
 
     const userInfo = jsonUserInfo.response;
+
```
Was this an intentional change or an accidental space?

lhao03 replied:
accidental LOL. i always put spaces when i shouldnt hahah

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -39,9 +40,10 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
 
   public componentDidUpdate = (prevProps: IPermissionProps) => {
-    if ((prevProps.sample !== this.props.sample) || (prevProps.permissionsPanelOpen !== this.props.permissionsPanelOpen)) {
+    if ((prevProps.sample !== this.props.sample) || (prevProps.permissionsPanelOpen !== this.props.permissionsPanelOpen) || (prevProps.permissionModeType !== this.props.permissionModeType)) {
```
```suggestion
    if (prevProps.sample !== this.props.sample 
            || prevProps.permissionsPanelOpen !== this.props.permissionsPanelOpen 
            || prevProps.permissionModeType !== this.props.permissionModeType) {
```
I think this line is a bit too long for general code cleanliness, since you have to scroll to read it.

Splitting the checks by the ||'s (or &&'s in other cases) is common practice to my knowledge and generally makes the code a lot cleaner/easier to read ^^

lhao03 replied:
yeah ill change that!

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -60,7 +63,8 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
       || nextProps.permissionsPanelOpen !== this.props.permissionsPanelOpen
-      || nextState.permissions !== this.state.permissions;
+      || nextState.permissions !== this.state.permissions
+      || nextProps.permissionModeType !== this.props.permissionModeType
```
```suggestion
      || nextProps.permissionModeType !== this.props.permissionModeType;
```
Line 63 had the semicolon but line 67 forgot it ^^

lhao03 replied:
(´＠＿＠｀ ) great catch!

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -169,43 +174,47 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
           isResizable: true,
-          minWidth: (tokenPresent) ? 400 : 600,
-          maxWidth: (tokenPresent) ? 600 : 1000,
+          minWidth: (tokenPresent) ? ((isCurrentPermTypeDelegated) ? 400 : 800) : 600,
+          maxWidth: (tokenPresent) ? ((isCurrentPermTypeDelegated) ? 600 : 1300) : 1000,
```
```suggestion
          minWidth: (tokenPresent) 
              ? (isCurrentPermTypeDelegated ? 400 : 800) 
              : 600,
          maxWidth: (tokenPresent) 
              ? (isCurrentPermTypeDelegated ? 600 : 1300) 
              : 1000,
```
I think that 2 ternary operators in a row becomes a bit more difficult to notice unless you distinguish them; With this change, I just made the first ternary more apparent, and kept the second ternary the same.

Also, I feel like having the parenthesis around `tokenPresent` makes sense visually because it is an indicator that it is a boolean for the ternary, but since `isCurrentPermTypeDelegated` is already a fully parenthesis-encapsulated ternary, having parenthesis around it feels more difficult to read.

lhao03 replied:
agreed!

jerryxihe replied:
Thanks for the suggestion @AyuAi, we weren't really sure what to do with this!

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -169,43 +174,47 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
       }
     ];
 
+    const isCurrentPermTypeDelegated = permissionModeType === DISPLAY_DELEGATED_PERMISSIONS;
```
I feel like the naming of this variable is a bit weird, since `tokenPresent` serves a similar purpose of describing the state but is worded differently. Maybe `isCurrentPermTypeDelegated` -> `delegatedPermissions` serves to more concisely describe the state in the same manner as `tokenPresent` does?

lhao03 replied:
yeah I agree!

jerryxihe replied:
https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/12#discussion_r663284122 Here's the comment @ChangshengYan made on the original PR. I think the idea was to make the variable identifiable as a boolean from the name alone. If you think `delegatedPermissions` fits that criteria, then I'm on board for the change.

csyan5 replied:
I feel it's weird to compare two Booleans and assign it to anther Boolean. Had a comment above regarding making `permissionModeType` as an enum. Then `DISPLAY_DELEGATED_PERMISSIONS` and `DISPLAY_APPLICATION_PERMISSIONS` should be able to be replaced by the enum? (Correct me if I'm wrong lol)

ayubun replied:
@jerryxihe You could even make it `isDelegatedPermissions` actually if you want extra clarity; It doesn’t necessarily have to match the older codebase items since refactoring stuff is pretty normal

jerryxihe replied:
c99a5c534eba1492284a1e3e4a9f34c2193d5eb8 changed to `permissionModeTypeIsUser`

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -169,43 +174,47 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
-          maxWidth: 100
+          key: 'isAdmin',
+          isResizable: true,
+          name: messages['Admin consent required'],
```
```suggestion
          name: translateMessage('Admin consent required'),
```
See https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/17#pullrequestreview-699168629 for a more detailed explanation of this

omersayshi commented:
I think this should be put in the drafts, as application token retrieval is still under the works so the feature is not complete

ayubun commented on src/app/views/query-runner/request/permissions/TabList.tsx:
```diff
@@ -52,25 +62,27 @@ const TabList = ({ columns, classes, renderItemColumn, renderDetailsHeader, maxH
-        {!tokenPresent && <FormattedMessage id='sign in to consent to permissions' />}
-        {tokenPresent && <FormattedMessage id='permissions required to run the query' />}
+        {!tokenPresent && <FormattedMessage id={permissionModeType ? 'sign in to consent to permissions' : 'sign in to consent to application permissions'} />}
+        {tokenPresent && <FormattedMessage id={permissionModeType ? 'permissions required to run the query' : 'application permissions required to run the query'} />}
```
```suggestion
        {!tokenPresent 
            && <FormattedMessage id={permissionModeType 
                ? 'sign in to consent to permissions' 
                : 'sign in to consent to application permissions'} />}
        {tokenPresent 
            && <FormattedMessage id={permissionModeType 
                ? 'permissions required to run the query' 
                : 'application permissions required to run the query'} />}
```
I think these lines started to get a bit too long without breaks.

ayubun commented on src/app/views/settings/Settings.tsx:
```diff
@@ -81,6 +82,14 @@ function Settings(props: ISettingsProps) {
       menuItems.push(
+        {
+          key: 'switch-user-app-mode',
+          text: messages[permissionModeType ? "Use Explorer as sample Teams application" : "Use Explorer as logged-in user"],
```
```suggestion
          text: translateMessage(permissionModeType 
              ? "Use Explorer as sample Teams application" 
              : "Use Explorer as logged-in user"),
```
Switched to `translateMessage()` + Linebreaks for cleanliness

ayubun commented on src/app/services/actions/permissions-action-creator.ts:
```diff
@@ -43,8 +47,9 @@ export function fetchScopes(): Function {
-      const { devxApi, permissionsPanelOpen, sampleQuery: query }: IRootState = getState();
+      const { devxApi, permissionsPanelOpen, permissionModeType, sampleQuery: query }: IRootState = getState();
       let permissionsUrl = `${devxApi.baseUrl}/permissions`;
+      const scope = permissionModeType === DISPLAY_DELEGATED_PERMISSIONS ? PERMS_SCOPE.WORK : PERMS_SCOPE.APPLICATION;
```
```suggestion
      const scope = permissionModeType === (DISPLAY_DELEGATED_PERMISSIONS 
          ? PERMS_SCOPE.WORK 
          : PERMS_SCOPE.APPLICATION);
```
This line feels a bit weird without a linebreak, I think

ayubun commented on src/app/views/App.tsx:
```diff
@@ -351,13 +352,13 @@ class App extends Component<IAppProps, IAppState> {
                   minimised,
-                  this.toggleSidebar
-                )}
+                  this.toggleSidebar)}
```
Is there a reason for this? I would personally argue the code was cleaner/easier to read before

lhao03 replied:
i'm not exactly sure what you are referring to 😅

ayubun replied:
This commit section takes a 2 line and just removes a newline so that it becomes a one-liner.

The problem though is that there are lines above which show it is a multiline object, so I just felt that this commit was unnecessary.

ayubun commented on src/app/views/authentication/profile/Profile.tsx:
```diff
@@ -3,7 +3,6 @@ import React, { Component } from 'react';
 import { injectIntl } from 'react-intl';
 import { connect } from 'react-redux';
 import { bindActionCreators, Dispatch } from 'redux';
-
```
This linebreak looks intentionally placed to separate the modules from the files in the import list. I don't fully agree with removing it.

lhao03 replied:
yeah i can put that back

ayubun commented on src/app/views/authentication/profile/Profile.tsx:
```diff
@@ -106,7 +106,7 @@ export class Profile extends Component<IProfileProps, IProfileState> {
       imageUrl: user.profileImageUrl,
       imageInitials: this.getInitials(user.displayName),
-      text: user.displayName,
+      text: user.displayName + ' ' + (this.props.permissionModeType ? messages['As user'] : messages['As Teams app']),
```
```suggestion
      text: user.displayName + ' ' 
          + (this.props.permissionModeType 
              ? translateMessage('As user') 
              : translateMessage('As Teams app')),
```
This line felt annoying to read without any breaks;
Linebreaks for clarity, plus the change from `messages[]` to `translateMessage()`

ayubun commented on src/app/views/query-runner/request/Request.tsx:
```diff
@@ -10,7 +10,6 @@ import React, { Component, CSSProperties } from 'react';
 import { injectIntl } from 'react-intl';
 import { connect } from 'react-redux';
 import { bindActionCreators, Dispatch } from 'redux';
-
```
Same comment as before; I don't agree with this line removal 👀

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -169,43 +174,47 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
+          fieldName: 'isAdmin',
+          minWidth: (tokenPresent) ? 150 : 200,
+          maxWidth: (tokenPresent) ? 200 : 300,
+          ariaLabel: translateMessage('Administrator permission')
```
Noice usage of translateMessage :D

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -50,6 +52,7 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
     }
   }
 
+
```
```suggestion
```
Accidental newline?

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -273,14 +282,15 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
 }
 
-function mapStateToProps({ sampleQuery, scopes, authToken, consentedScopes, dimensions, permissionsPanelOpen }: IRootState) {
+function mapStateToProps({ sampleQuery, scopes, authToken, consentedScopes, dimensions, permissionsPanelOpen, permissionModeType }: IRootState) {
```
This line is getting a tiny bit long, in my opinion; I am actually unfamiliar with TS/JS checkstyle commonalities for functions so I don't know how to split it personally

csyan5 replied:
Could be something like 
```suggestion
function mapStateToProps({ 
    sampleQuery, 
    scopes, 
    authToken, 
    consentedScopes, 
    dimensions, 
    permissionsPanelOpen, 
    permissionModeType }: IRootState) {
```

ayubun replied:
Thanks for the suggestion! I think that looks a lot better 😄

ayubun replied:
Resolving in #23 - Review when you can!

ayubun commented on src/app/views/query-runner/request/permissions/TabList.tsx:
```diff
@@ -52,25 +62,27 @@ const TabList = ({ columns, classes, renderItemColumn, renderDetailsHeader, maxH
         layoutMode={DetailsListLayoutMode.justified}
         onRenderDetailsHeader={(props?: any, defaultRender?: any) => renderDetailsHeader(props, defaultRender)} />
-      {permissions && permissions.length === 0 &&
+      {filteredPermissions && filteredPermissions.length === 0 &&
         displayNoPermissionsFoundMessage()
```
```suggestion
      {filteredPermissions 
        && filteredPermissions.length === 0 
        && displayNoPermissionsFoundMessage()
```
This isn't on ya'll at all because this line existed before, but I'm pretty sure the &&'s should be newline-d here for readability.

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/sidebar/Sidebar.tsx:
```diff
@@ -5,7 +5,6 @@ import { telemetry } from '../../../telemetry';
 import History from './history/History';
 import SampleQueries from './sample-queries/SampleQueries';
 
-
```
Reason for this? 👀 I think double-space is an intended separator here

lhao03 replied:
tbh I have no idea（´＠＿＠｀ ）

csyan5 replied:
fyi that the Azure Portal codebase I'm working on has tslint checking against double empty lines (and many other rules..)

lhao03 replied:
resolved in https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/commits/55666db4d0964581599d5f907b355fb066034d67

ayubun commented on src/app/views/query-runner/request/permissions/Permission.tsx:
```diff
@@ -169,43 +174,47 @@ export class Permission extends Component<IPermissionProps, IPermissionState> {
+        columns.push(
+          {
+            key: 'consented',
+            name: messages.Status,
```
I'm actually not entirely sure how to properly localize this; Would you simply do `translateMessage(messages.Status)`? I would need to test it out, but I am not familiar enough to pinpoint quickly where this change would affect

ayubun reviewed (changes requested):
For all the localization comments, check out https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/17#pullrequestreview-699168629 👀 

For the rest, I think the comments are sufficient to explain, but feel free to poke me about any of them ☜(ﾟヮﾟ☜)

Despite all of the comments I made, awesome work on the PR guys 🙏 🙏

lhao03 commented on src/app/services/actions/permissions-action-creator.ts:
```diff
@@ -13,6 +12,11 @@ import {
+import {
+  PERMS_SCOPE,
+  DISPLAY_DELEGATED_PERMISSIONS,
+  DISPLAY_APPLICATION_PERMISSIONS
```
```suggestion
  DISPLAY_APPLICATION_PERMISSIONS;
```

lhao03 committed:
Update src/app/services/actions/permission-mode-action-creator.ts

lhao03 committed:
Update src/app/services/actions/permission-mode-action-creator.ts

lhao03 commented:
![image](https://user-images.githubusercontent.com/55033656/124515937-794eb900-dd95-11eb-9ee3-9ea7afb763d1.png)
sadge

lhao03 commented:
Changes based on the code suggestions can be seen in this PR: https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/20

lhao03 committed:
Merge pull request #18 from LokiLabs/interns/lhao03/38301-panel-rsc-delegated

AB#38301 Make panel permissions switch between delegated and RSC

csyan5 commented on src/app/views/App.tsx:
```diff
@@ -374,7 +375,7 @@ class App extends Component<IAppProps, IAppState> {
               {displayContent && <>
                 <div style={{ marginBottom: 8 }}>
-                  <QueryRunner onSelectVerb={this.handleSelectVerb} />
+                  <QueryRunner onSelectVerb={this.handleSelectVerb} permissionModeType={permissionModeType} />
```
looks like `permissionModeType` is passed around in `QueryRunner.tsx` and `Request.tsx` but only used in `Permission.tsx`. I believe we can remove the references in `QueryRunner.tsx` and `Request.tsx`?

lhao03 replied:
https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/27

csyan5 commented on src/app/views/authentication/profile/Profile.tsx:
```diff
@@ -106,7 +106,7 @@ export class Profile extends Component<IProfileProps, IProfileState> {
       imageUrl: user.profileImageUrl,
       imageInitials: this.getInitials(user.displayName),
-      text: user.displayName,
+      text: user.displayName + ' ' + (this.props.permissionModeType ? messages['As user'] : messages['As Teams app']),
```
I would suggest making `permissionModeType` as an Enum and add maps for conditional display for readability and maintainability. 
```
const enum PermissionModeType {
  User,
  TeamsApp,
}

 const permissionModeTypeDisplayName = {
      [PermissionModeType.User]: 'As user',
      [PermissionModeType.TeamsApp]: 'As Teams app',
    }
```
then you can use
```
text: user.displayName + ' ' + translateMessage(permissionModeTypeDisplayName[this.props.permissionModeType])
```

ayubun replied:
Making a branch to test this out; Thank you for the suggestion!

csyan5 commented on src/types/request.ts:
```diff
@@ -31,6 +31,7 @@ export interface IRequestComponent {
     actions: {
         setDimensions: Function;
     };
+    permissionModeType: boolean;
```
Please see one of my comments above. Then we could remove `permissionModeType` here?

jerryxihe replied:
75764e625bcc3e45ab338fb1ad0d5b0e24f5fd94

csyan5 reviewed (dismissed):
Left some comments and please let me know if you have any feedback! 😃

lhao03 committed:
MVP fixes for #16 (#20)

* implemented changes from comments

* Update src/app/views/query-runner/request/permissions/Permission.tsx

Co-authored-by: Ayu <ayu@ayu.dev>

ayubun committed:
Resolve https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/files#r664175395

ayubun committed:
Resolve https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16/files#r663967497

ayubun committed:
Localization fixes

ayubun committed:
Change permissionModeType to enum
(Resolve https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16#discussion_r664172252)

ayubun committed:
Update tests

ayubun committed:
Remove unused import

jerryxihe committed:
Re-ordered PERMISSION_MODE_TYPE enum to fix wrong mode bug

ayubun committed:
Apply suggestions from code review

Comment updates from @lhao03

Co-authored-by: Lucy Hao <55033656+lhao03@users.noreply.github.com>

jerryxihe committed:
Added error message when viewing `Access token` tab in app mode

ayubun committed:
Resolve @ChangshengYan's code review for PR#23

jerryxihe committed:
Changed remaining `permissionModeType` booleans to enums

ayubun committed:
Merge pull request #23 from LokiLabs/interns/AyuAi/PR#16-changes

PR#16 Changes Branch

jerryxihe committed:
Moved app mode access token error message to GE.json

jerryxihe committed:
Merge pull request #26 from LokiLabs/interns/jerryxihe/38717-access-token-error-in-app-mode

AB#38717 Interns/jerryxihe/38717 access token error in app mode

lhao03 committed:
AB#38767 remove unused props (#27)

* remove unused prosp

* Removed remaining unused `permissionModeType` and `changeMode()` instances

* remove unused prosp

Co-authored-by: Jerry He <jerryxhe@gmail.com>

lhao03 committed:
AB#38239-switch-user-profile-to-graph-explorer-sample-app (#21)

* implemented changes from comments

* switch user profile name/image when in app mode

* changes from PR comments

* remove unused prosp

* switch user profile name/image when in app mode

* changes from PR comments

ayubun commented:
This is not necessary to resolve prior to MVP, but for the sake of localization, I am unsure on this https://github.com/LokiLabs/microsoft-graph-explorer-v4/pull/16#discussion_r663985222

Maybe it is fine as-is?

acchiang commented:
poked around locally on this branch, looks good!
